### DISCUSSION TITLE
Minor feature request: Add a way to specify how `Serializer.to_repr` handles `None

### DISCUSSION THREAD

**Discussion by alexlambson** [encode/django-rest-framework#7811] [Ideas & Suggestions]
Currently all None values are sent to the API as None. It would be preferable to tell drf what to use for "None". e.g. our UI currently expects an empty list for a field if it's none because we were using a SerializerMethodField. We switched to a custom field to allow for read-write, and now DRF returns None even though we need an empty list. 

I propose adding a way to bypass this `if` and always call `to_representation` regardless of `attribute`, or add a way to specify what a field's `"None"` is. I am open to writing this myself if the team wants it, but doesn't want to write it. 

```
# in rest_framework.serializers.Serializer.to_representation.

check_for_none = attribute.pk if isinstance(attribute, PKOnlyObject) else attribute
if check_for_none is None:
    ret[field.field_name] = None
else:
    ret[field.field_name] = field.to_representation(attribute)
```

Could maybe be changed to something like this

```
check_for_none = attribute.pk if isinstance(attribute, PKOnlyObject) else attribute
if check_for_none is None:
    ret[field.field_name] = field.none_value  # defaults to "None"
else:
    ret[field.field_name] = field.to_representation(attribute)
```

## Checklist

- [x] I have verified that that issue exists against the `master` branch of Django REST framework.
- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [x] This is not a usage question. (Those should be directed to the [discussion group](https://groups.google.com/forum/#!forum/django-rest-framework) instead.)
- [x] This cannot be dealt with as a third party library. (We prefer new functionality to be [in the form of third party libraries](https://www.django-rest-framework.org/community/third-party-packages/#about-third-party-packages) where possible.)
- [x] I have reduced the issue to the simplest possible case.
- [ ] I have included a failing test as a pull request. (If you are unable to do so we can still accept the issue.)

**Comment by alexlambson** [comment]
Just following up that I am willing to write this should the team want to add this functionaility to DRF.

**Reply by lovelydinosaur** [reply]
Thanks. I don't think we particularly want functionality to serialize `None` values as something other than `None` outputs. It's not something that's come up as a request, and it's hard to see good use-cases for it. If it *is* something that you need in a project, I'd suggest a custom serializer class, with a slightly adapted `to_representation` method based on the existing behaviour.

**Reply by alexlambson** [reply]
That is how we ended up getting around it. But if you guys don't really see a use for this then I am fine with closing the discussion.

**Reply by jprestel-rue** [reply]
We just ran into this building out an API. We have a few object classes where fields that aren't specified on the object are being serialized as None. We found it unintuitive that specifying a `default` value for these fields had no effect. In retrospect, I see the design choice that the framework wants to treat a `None` as an explicit, intentional, meaningful `None`. Strange how I'm longing for JavaScript's concept of `undefined`.

In any case, just wanted to add that we definitely had a use case for being able to handle `None` in a custom way for specific fields.

Since we want this on a per-field basis, the workaround we're considering is to just modify the object to set the defaults before it ever hits the serializer.

**Comment by pbav** [comment]
I just run into this issue when I wanted a nullable boolean field to represent None as False:
`attr = serializers.BooleanField(read_only=True, allow_null=False)`

The serializers.BooleanField's to_representation is:
```
    def to_representation(self, value):
        if value in self.TRUE_VALUES:
            return True
        elif value in self.FALSE_VALUES:
            return False
        if value in self.NULL_VALUES and self.allow_null:
            return None
        return bool(value)
```
It's clearly supposed to return False when `self.allow_null` is False.
However now it is never called because of:
```
class Serializer(BaseSerializer, metaclass=SerializerMetaclass):
...
    def to_representation(self, instance):
...
            # We skip `to_representation` for `None` values so that fields do
            # not have to explicitly deal with that case.
            #
            # For related fields with `use_pk_only_optimization` we need to
            # resolve the pk value.
            check_for_none = attribute.pk if isinstance(attribute, PKOnlyObject) else attribute
            if check_for_none is None:
                ret[field.field_name] = None
            else:
                ret[field.field_name] = field.to_representation(attribute)
```

**Comment by ckkz-it** [comment]
I've just run exactly into the same issue with the `BooleanField` that @pbav had. Want to tag @tomchristie and @alexlambson for visibility. Looks like it's been there for years.
